### PR TITLE
Email fixture

### PR DESCRIPTION
**Pull request by roydekleijn**
E-mail fixture to be able to read mails and extract text

### CODE REVIEW

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * | set mail provider with host | <i>host</i> | port | <i>port</i> | user | <i>username</i> | password | <i>password</i> |
+     */
+    public void setMailProviderWithHostPortUserPassword(final String host, final String port, final String username, final String password) {
+        final Properties props = new Properties();
+        props.setProperty("mail.smtp.host", host);
+        props.setProperty("mail.smtp.socketFactory.port", port);
+        props.setProperty("mail.smtp.socketFactory.class", "javax.net.ssl.SSLSocketFactory");
+        props.setProperty("mail.smtp.auth", "true");
+        props.setProperty("mail.smtp.port", port);
```
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: smtp is for sending mail. Should this not be pop/imap(s) instead?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        try {
+            final Session session = Session.getDefaultInstance(props, null);
+            store = session.getStore("imaps");
+            store.connect("smtp.gmail.com", username, password);
```
Comment (tcnh) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Hardcoded for gmail smtp?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java (hunk 1 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: correct!

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java (hunk 2 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: correct, changed it..

**Comment by roydekleijn** [comment]
It can be used like this

image

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+     * | set mail provider with host | <i>host</i> | port | <i>port</i> | user | <i>username</i> | password | <i>password</i> |
+     */
+    public void setMailProviderWithHostPortUserPassword(final String host, final String port, final String username, final String password) {
+        final Properties props = new Properties();
+        try {
+            final Session session = Session.getDefaultInstance(props, null);
+            store = session.getStore("imaps");
+            store.connect(host, username, password);
+        } catch (final MessagingException e) {
+            throw new StopTestException("Cannot connect to mailserver");
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Please include `e` as cause

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            Message msg = getRecentMessagesMatching(inbox, params);
+            return getBody(msg);
+        } catch (final Exception ex) {
+            throw new SlimFixtureException(false, "No message found with search params: " + params.toString(), ex);
+        }
+    }
+
+    private Folder getInboxFolder() throws MessagingException {
+        final Folder inbox = store.getFolder("inbox");
+        inbox.open(Folder.READ_WRITE);
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Do you need write permission?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        public boolean isFinished() {
+            return mails != null && mails.size() > 0;
+        }
+
+        @Override
+        public void repeat() {
+            try {
+                mails = getMessagesMatching(inbox, params);
+            } catch (RuntimeException ex) {
+            }
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Shouldn't th exception at least be logged (at debug or something). 
If ignoring is OK, please provide comment why.

File: src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeIOException.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package nl.hsac.fitnesse.fixture.slim.exceptions;
+
+import java.io.IOException;
+
+/**
+ * Created by roydekleijn on 04/12/2017.
+ */
+public class RuntimeIOException extends RuntimeException {
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeIOException.java]: Just throw runtimeexceptions whith IOException as cause

File: src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeMessageException.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package nl.hsac.fitnesse.fixture.slim.exceptions;
+
+public class RuntimeMessageException extends RuntimeException {
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeMessageException.java]: Just throw runtimeexceptions whith Exception as cause

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        try {
+            for (final Message mail : mails) {
+                if (mail.getSentDate().after(HOURS_BACK)) {
+                    return mail;
+                }
+            }
+        } catch (MessagingException ex) {
+            throw new RuntimeMessageException(ex);
+        }
+        throw new RuntimeIOException();
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Why not throw a SlimFixtureException stating that no messages were present sent after HOURS_BACK?
This provides more background about the limit used at runtime.

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        }
+        throw new SlimFixtureException(false, format("No match found in text %s with regex %s", text, regexpattern));
+    }
+
+    private String getMailText(SearchParameters params) {
+        try {
+            final Folder inbox = getInboxFolder();
+            Message msg = getRecentMessagesMatching(inbox, params);
+            return getBody(msg);
+        } catch (final Exception ex) {
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: If a SlimFixtureException is thrown I would not create a new one, but just rethrow the existing

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        }
+    }
+
+    /**
+     * | ensure | mail received with subject | <i>subject</i> | from | <i>receiver</i> |
+     *
+     * @param subject
+     * @return
+     */
+    public String mailReceivedWithSubjectFrom(final String subject, final String receiver) {
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: a mail received from, I would expect the parameter to be the sender, not the reciever....? 

But given the code I see below I expect this method should probably be called: `mailReceivedByWithSubject` and the order of the parameters reversed. (Given that the receiver is likely to be shorter it is also good practice to have that as first parameter, to make rows easier to read)

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        }
+    }
+
+    private String tryBody(Message msg) throws IOException, MessagingException {
+        String message = "";
+        if (msg != null && msg.getContent() instanceof MimeMultipart) {
+            final Multipart multipart = (Multipart) msg.getContent();
+            message = IOUtils.toString(multipart.getBodyPart(0).getInputStream());
+        }
+        if (message != null) {
```
Comment (fhoeben) [inline src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: Should this not be `if ("".equals(message) && msg != null) {`?

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java (hunk 4 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: changed

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    /**
+     * | $result= | extract text with | <i>regex</i> | from mail with | <i>subject</i> | from | <i>receiver</i> |
+     */
+    public String extractTextWithFromMailWithFrom(final String regex, final String subject, final String receiver) {
+        final String text = getMailText(new SearchParameters(subject, receiver, HOURS_BACK));
+        return getText(text, regex);
+    }
+
+    private String getText(final String text, final String regexpattern) {
+        LOGGER.debug("Text found: " + text);
```
[the remark this answers is not in the document]
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: done

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+import java.util.regex.Pattern;
+
+import static java.lang.String.format;
+
+public class EmailFixture extends SlimFixture {
+    private final static Logger LOGGER = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+    private static final Date HOURS_BACK = new Date(System.currentTimeMillis() - 1000 * 60 * 60 * 24);
+    private Store store;
+
+    public EmailFixture() {
```
[the remark this answers is not in the document]
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: done

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+    }
+
+    /**
+     * | set mail provider with host | <i>host</i> | port | <i>port</i> | user | <i>username</i> | password | <i>password</i> |
```
[the remark this answers is not in the document]
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: was used before... removed now

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+     */
+    public void setMailProviderWithHostPortUserPassword(final String host, final String port, final String username, final String password) {
+        final Properties props = new Properties();
```
[the remark this answers is not in the document]
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: changed it

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java (hunk 3 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: done

File: src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java (hunk 10 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java]: good one! Like this we can easily use email aliases

File: src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/CouldNotFindMessageException.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package nl.hsac.fitnesse.fixture.slim.exceptions;
+
+import nl.hsac.fitnesse.fixture.slim.EmailFixture;
+
+import javax.mail.Folder;
+
+import static java.lang.String.format;
+
+public class CouldNotFindMessageException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/CouldNotFindMessageException.java]: done

File: src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeIOException.java (hunk 6 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeIOException.java]: done

File: src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeMessageException.java (hunk 7 above)
Comment (roydekleijn) [reply src/main/java/nl/hsac/fitnesse/fixture/slim/exceptions/RuntimeMessageException.java]: done

**Comment by fhoeben** [comment]
I pushed some code cleanup, can you check I didn't break anything?

I think the fixture could be made a bit more 'extension friendly' allowing users to customise it using a subclass. Can you look at making some more methods protected and see whether some restructuring could aid this.

**Comment by fhoeben** [comment]
I mean that the class is currently coded quite specifically to retrieve messages sent/received after a certain time in an Inbox folder sent to a known receiver via imaps. I feel this will probably not ininisuite everyones needs.
At the moment, with most things private, this class can not easily be extended to make different choices. I would like to see it refactored to ensure it also helps people that have different needs when testing mails sent.

The current state of the class feels a bit 'unfinished' I don't think it can be included in this project in its current state. I feel it would be better to have it in a separate project/library (for now), with version below 1.0.0. Similar to what I did with [PdfFixture](https://github.com/fhoeben/hsac-fitnesse-pdf). 
Of course it can still be published to Maven Central, so using it would just be a matter of adding a dependency.

I don't mind setting up a new project/repo for that, and publishing to Maven Central. But you could easily do that yourself also, of course. What would you prefer?

**Comment by roydekleijn** [comment]
Ok, will close this PR.
However, the most trivial scenario that can be tested is forgot password functionally, extract code/link from mail and reset password.

**Comment by fhoeben** [comment]
I will setup a repo

**Comment by fhoeben** [comment]
Moved to https://github.com/fhoeben/hsac-fitnesse-email

**Comment by fhoeben** [comment]
Release 0.0.1 has been published to Maven Central

**Pull request closed without merging**